morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs

```diff
@@ -0,0 +1,397 @@
+import Ouroboros.Network.TxSubmission.Inbound (TraceTxSubmissionInbound)
+import Ouroboros.Network.TxSubmission.Outbound (TraceTxSubmissionOutbound)
+import Prettyprinter (Doc, Pretty (..), list, viaShow, (<+>))
+import Prelude (String, id)
+
+class MPretty a where
+  mpretty :: a -> Doc ann
+
+instance (HashAlgorithm h, BftCrypto c) => MPretty (MorphoBlock h c) where
+  mpretty MorphoBlock {..} =
+    "("
+      <> mpretty morphoPrev
+      <> "->"
+      <> mpretty morphoHeaderHash
+      <> ","
+      <> mpretty morphoSlotNo
+      <> ","
+      <> mpretty morphoTxs
+      <> ")"
+    where
+      MorphoHeader {..} = morphoHeader
+      MorphoStdHeader {..} = morphoHeaderStd
+      MorphoBody {..} = morphoBody
+
+instance
+  (HashAlgorithm h, BftCrypto c) =>
+  MPretty (Header (MorphoBlock h c))
+  where
+  mpretty MorphoHeader {..} =
+    "("
+      <> mpretty morphoPrev
+      <> "->"
+      <> mpretty morphoHeaderHash
+      <> ","
+      <> mpretty morphoSlotNo
+      <> ")"
+    where
+      MorphoStdHeader {..} = morphoHeaderStd
+
+instance MPretty MorphoBlockTx where
+  mpretty = mpretty . morphoBlockGenTx
+
+instance MPretty Tx where
+  mpretty (Tx vote) = mpretty vote
+
+instance MPretty Vote where
+  mpretty = viaShow
+
+instance MPretty (Hash h a) where
+  mpretty = pretty . Text.take 8 . hashToTextAsHex
+
+instance MPretty SlotNo where
+  mpretty (SlotNo n) = pretty n
+
+instance MPretty BlockNo where
+  mpretty (BlockNo n) = pretty n
+
+instance MPretty (TraceBlockchainTimeEvent t) where
+  mpretty (TraceStartTimeInTheFuture (SystemStart start) toWait) =
+    "Waiting"
+      <+> viaShow toWait
+      <+> "until genesis start time at"
+      <+> viaShow start
+  mpretty TraceSystemClockMovedBack {} =
+    "System clock moved back an acceptable time span"
+  mpretty TraceCurrentSlotUnknown {} = "Current slot is not yet known"
+
+instance
+  (HashAlgorithm h, BftCrypto c, Signable (BftDSIGN c) (MorphoStdHeader h c), MorphoStateDefaultConstraints h c) =>
+  MPretty (TraceChainSyncClientEvent (MorphoBlock h c))
+  where
+  mpretty = viaShow
+
+instance MPretty (TraceLocalTxSubmissionServerEvent (MorphoBlock h s)) where
+  mpretty = viaShow
+
+instance
+  ( Signable (BftDSIGN s) (MorphoStdHeader h s),
+    MorphoStateDefaultConstraints h s
+  ) =>
+  MPretty (ChainDB.TraceAddBlockEvent (MorphoBlock h s))
+  where
+  mpretty = viaShow
+
+instance (StandardHash blk) => MPretty (TraceChainSyncServerEvent blk) where
+  mpretty = viaShow
+
+instance MPretty (TraceBlockFetchServerEvent blk) where
+  mpretty = viaShow
+
+instance MPretty a => MPretty (TraceLabelCreds a) where
+  mpretty (TraceLabelCreds _ a) = mpretty a
+
+instance
+  ( Signable (BftDSIGN c) (MorphoStdHeader h c),
+    MorphoStateDefaultConstraints h c
+  ) =>
+  MPretty (TraceForgeEvent (MorphoBlock h c))
+  where
+  mpretty (TraceAdoptedBlock slotNo blk txs) =
+    "Adopted forged block for slot"
+      <+> pretty (unSlotNo slotNo)
```

Comment:
All of `pretty (unSlotNo slotNo)` can be replaced with `mpretty slotNo`, since `MPretty SlotNo` exists